<a href="https://colab.research.google.com/github/faisu6339-glitch/Deep-Learning/blob/main/forward_Progation_in_MLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### What is Forward Propagation in a Multi-Layer Perceptron (MLP)?

Forward propagation (also known as forward pass) is the process by which a neural network takes input data, processes it through its layers, and generates an output. In an MLP, this involves a series of matrix multiplications and activation functions.

Let's break down the process:

1.  **Input Layer**: The raw input data enters the network. Each input feature corresponds to a neuron in the input layer.

2.  **Hidden Layers**: The input then travels through one or more hidden layers. For each neuron in a hidden layer, two main operations occur:
    *   **Weighted Sum**: The inputs from the previous layer are multiplied by their respective weights, and a bias term is added. This forms a weighted sum (or 'net input'). Mathematically, for a neuron `j` in a hidden layer, the weighted sum `z_j` is calculated as:
        `z_j = Σ (w_ij * x_i) + b_j`
        where `x_i` are the inputs from the previous layer, `w_ij` are the weights connecting input `i` to neuron `j`, and `b_j` is the bias for neuron `j`.
    *   **Activation Function**: The weighted sum `z_j` is then passed through an activation function (e.g., ReLU, Sigmoid, Tanh). This introduces non-linearity into the network, allowing it to learn complex patterns. The output of the activation function `a_j` becomes the input for the next layer:
        `a_j = activation(z_j)`

3.  **Output Layer**: The process is repeated for the output layer. The outputs from the last hidden layer serve as inputs to the output layer. The output layer's activation function depends on the type of problem (e.g., softmax for multi-class classification, sigmoid for binary classification, linear for regression).

In essence, forward propagation is the flow of information from the input to the output, transforming the data at each step to extract relevant features and make predictions.

### Simple Python Code Example for Forward Propagation

In [1]:
import numpy as np

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def relu(x):
    return np.maximum(0, x)

# Define the input
x = np.array([0.5, 0.8]) # 2 input features

# Define weights and biases for a simple MLP with one hidden layer and one output layer

# Hidden Layer 1 (2 neurons)
# Each input feature (x_0, x_1) connects to each hidden neuron (h_0, h_1)
weights_h1 = np.array([
    [0.1, 0.4], # Weights for h_0 from x_0, x_1
    [0.3, 0.2]  # Weights for h_1 from x_0, x_1
])
bias_h1 = np.array([0.1, 0.2]) # Biases for h_0, h_1

# Output Layer (1 neuron for a simple regression or binary classification output)
# Each hidden neuron (h_0, h_1) connects to the output neuron (o_0)
weights_out = np.array([
    [0.6], # Weight for o_0 from h_0
    [0.7]  # Weight for o_0 from h_1
])
bias_out = np.array([0.3]) # Bias for o_0

print("Input x:", x)

# --- Forward Propagation ---

# 1. Calculate weighted sum for Hidden Layer 1
# z_h1 = x @ weights_h1.T + bias_h1 # If weights_h1 has shape (num_neurons, num_inputs)
# A common convention is (num_inputs, num_neurons) for weights, then x @ weights_h1
z_h1 = np.dot(x, weights_h1) + bias_h1
print("\nWeighted sum for Hidden Layer 1 (z_h1):", z_h1)

# 2. Apply activation function (e.g., ReLU) to Hidden Layer 1
a_h1 = relu(z_h1)
print("Output of Hidden Layer 1 after ReLU (a_h1):", a_h1)

# 3. Calculate weighted sum for Output Layer
z_out = np.dot(a_h1, weights_out) + bias_out
print("\nWeighted sum for Output Layer (z_out):", z_out)

# 4. Apply activation function (e.g., Sigmoid for binary classification or linear for regression) to Output Layer
# For this example, let's assume a linear output for simplicity, or sigmoid if it's a probability
# output = sigmoid(z_out) # For binary classification
output = z_out # For a linear output/regression
print("Final Output (after linear activation for simplicity):", output)

Input x: [0.5 0.8]

Weighted sum for Hidden Layer 1 (z_h1): [0.39 0.56]
Output of Hidden Layer 1 after ReLU (a_h1): [0.39 0.56]

Weighted sum for Output Layer (z_out): [0.926]
Final Output (after linear activation for simplicity): [0.926]


Input x: [0.5 0.8] - This is our initial input to the network.

Weighted sum for Hidden Layer 1 (z_h1): [0.39 0.56] - This is calculated by multiplying the input x by the weights_h1 and adding bias_h1. Each value corresponds to the net input for each neuron in the hidden layer.

Output of Hidden Layer 1 after ReLU (a_h1): [0.39 0.56] - The ReLU activation function was applied to z_h1. Since both values in z_h1 are positive, they remain unchanged by ReLU.

Weighted sum for Output Layer (z_out): [0.926] - This is the net input for the output neuron, calculated by multiplying a_h1 by weights_out and adding bias_out.

Final Output: [0.926] - In this simplified example, a linear activation (or no activation) was applied to z_out, resulting in the final prediction.

In this example:
*   The input `x` (0.5, 0.8) is fed into the network.
*   It's transformed by the weights and biases of the first hidden layer.
*   The ReLU activation function is applied to introduce non-linearity.
*   The output of the hidden layer becomes the input for the output layer.
*   The output layer performs its own weighted sum and (in this simplified case) a linear activation to produce the final output.

In [2]:
import numpy as np

# Input
X = np.array([1, 2, 3])

# Weights and bias (example)
W1 = np.array([[0.2, 0.4, 0.6],
               [0.1, 0.3, 0.5]])
b1 = np.array([0.1, 0.2])

# Hidden layer
Z1 = np.dot(W1, X) + b1
A1 = np.maximum(0, Z1)  # ReLU

# Output layer
W2 = np.array([[0.7, 0.8]])
b2 = np.array([0.3])

Z2 = np.dot(W2, A1) + b2
A2 = 1 / (1 + np.exp(-Z2))  # Sigmoid

print("Output:", A2)

Output: [0.98593637]
